File: src/main/java/nz/co/searchwellington/tagging/PlaceAutoTagger.java
package nz.co.searchwellington.tagging;

import java.util.HashSet;
import java.util.Set;

import nz.co.searchwellington.model.Resource;
import nz.co.searchwellington.model.Tag;
import nz.co.searchwellington.repositories.TagDAO;

import org.apache.log4j.Logger;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Component;

@Component
public class PlaceAutoTagger {
    
	private static Logger log = Logger.getLogger(PlaceAutoTagger.class);
    
    final String PLACES_TAG_NAME = "places";
        
    private TagDAO tagDAO;
    
    @Autowired
    public PlaceAutoTagger(TagDAO tagDAO) {
        this.tagDAO = tagDAO;
    }
    
    public Set<Tag> suggestTags(Resource resource) {         
    	Set<Tag> suggestedTags = new HashSet<Tag>();
    	
        Tag placesTag = tagDAO.loadTagByName(PLACES_TAG_NAME);
        if (placesTag != null) {            
            for (Tag placeTag : placesTag.getChildren()) {                          
                if (checkForMatchingTag(resource, placeTag)) {
                	log.info("Suggesting tag '" + placeTag.getDisplayName() + "' for resource: " + resource.getName());
                	suggestedTags.add(placeTag);
                }
            }            
        } else {
            log.warn("Could not find places tag.");
        }
        return suggestedTags;        
    }

    private boolean checkForMatchingTag(Resource resource, Tag tag) {
        boolean headlineMatchesTag = resource.getName().toLowerCase().contains(tag.getDisplayName().toLowerCase());
        boolean bodyMatchesTag = resource.getDescription().toLowerCase().contains(tag.getDisplayName().toLowerCase());
        return headlineMatchesTag || bodyMatchesTag;        
    }

}

File: src/main/scala/nz/co/searchwellington/tagging/PlaceAutoTagger.scala
package nz.co.searchwellington.tagging

import nz.co.searchwellington.ReasonableWaits
import nz.co.searchwellington.model.{Resource, Tag}
import nz.co.searchwellington.repositories.mongo.MongoRepository
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.stereotype.Component

import scala.concurrent.{ExecutionContext, Future}

@Component class PlaceAutoTagger @Autowired() (val mongoRepository: MongoRepository) extends TagAncestors
  with ReasonableWaits {

  private final val PLACES_TAG_NAME = "places"

  def suggestTags(resource: Resource)(implicit ec: ExecutionContext): Future[Set[Tag]] = {

    def getAllPlaces: Future[Seq[Tag]] = {
      mongoRepository.getTagByUrlWords(PLACES_TAG_NAME).flatMap { maybePlaceTag =>
        maybePlaceTag.map(descendantsOf).getOrElse(Future.successful(Seq.empty))
      }
    }

    def checkForMatchingTag(resource: Resource, tag: Tag): Boolean = {
      val headlineMatchesTag = resource.title.toLowerCase.contains(tag.getDisplayName.toLowerCase)
      val bodyMatchesTag = resource.description.exists(d => d.toLowerCase.contains(tag.getDisplayName.toLowerCase))
      headlineMatchesTag || bodyMatchesTag
    }

    getAllPlaces.map { places =>
      places.filter(p => checkForMatchingTag(resource, p)).toSet
    }
  }

}
